### PR TITLE
Bring input setters on par with type getters

### PR DESCRIPTION
**Pull request by Lappihuan**
The goal of this PR is to bring setters on Input Annotated Objects on par with getters.
Since the introduction of Input Types in #269 setters were very simple and did not bring any Field Middleware functionality that getters have (Autowire, Security, Right, Assert etc..).

With this Update setters have their own InputFieldMiddleware system that is pretty much 1:1 to the FieldMiddleware system allowing Autowire, Security and Right annotations to work.
Assertation would need to be added separately into the respective repositories for Symfony and Laravel.

While this seems like a relatively small addition, this allows for a very potent pattern with Doctrine Entities.
I will explain this pattern in depth as a example but this PR brings many more possibilities not outlined here.

Entities can now not only have computed fields on getters but also on setters leveraging the power of Autowire to set up references for example:
```php
#[ORM\Entity]
#[GQL\Type(default: true)]
#[GQL\Input(name: "UpdateAccountInput", update: true, default: true)]
class Account {
//...
    #[ORM\ManyToOne]
    #[ORM\JoinColumn(name: "lang_id", referencedColumnName: "lang_id", nullable: false)]
    private Language $lang;
//...
    #[GQL\Field]
    public function setLang(LanguageEnum $lang, #[GQL\Autowire] EntityManagerInterface $entityManager): void
    {
        $this->lang = $entityManager->getReference(Language::class, $lang->getValue());
    }

    #[GQL\Field]
    public function getLang(): LanguageEnum
    {
        return LanguageEnum::from($this->lang->getLangId());
    }
}
```
This allows for a very clean separation of ORM responsibility around the private properties and the public API using the setter and getter functions without the need to map between DTOs and Entities, there is even potential for nested graph traversing in mutations but this would need to be figured out.

Additionally using Doctrines [Embeddables](https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/tutorials/embeddables.html#separating-concerns-using-embeddables) one can fully abstract the public data structure while still working with Doctrine Entities that can be persisted without any mapping.

Unrelated to this lib we even have a merge service that allows to persist partially hydrated Entities like this:
```php
    #[GQL\Mutation]
    public function updateAccount(
        ID $id,
        #[GQL\UseInputType(inputType: "UpdateAccountInput!")]
        Account $account
    ): Account
    {
        $account->setID($id);
        $merged = $this->mergeService->merge($account);
        $this->entityManager->flush();
        return $merge;
    }
```
The as for Validation directly on setters, i will make a PR on the [Symfony Validator Bridge](https://github.com/thecodingmachine/graphqlite-symfony-validator-bridge) Repo asap.

There is still documentation to be made around this feature and maybe some edge cases to be discussed.

@oojacoboo there are some commits left over from your input improvements since we initially based our branch on your.
Let me know if this needs cleanup or can be ignored.

Overall working on this PR gave us a lot of insights into how this lib is working and i already have a few more ideas on how to improve it :)

Looking forward to your feedback!

### CODE REVIEW

**Comment by oojacoboo** [comment]
@Lappihuan I'll try to have a deeper look at this in the next few days.  

I'm not a huge fan of reusing the Doctrine entities as input types, personally.  I know other people are doing this as well and I don't see why we shouldn't allow more support around this.  There are a few things I really don't like, however, the autowiring being a big one.  Adding a service as a param to your domain model setters is really not ideal.  It works, but ideally your domain would be more pure and you'd use services outside of the domain models.  Autowiring is really a nasty hack overall, even outside the domain IMO.  That said, it's already implemented and I'm not sure the cat gets put in the bag around it.  Further, my more pure preferences may not be shared by many others.

I also still think support for setters is beneficial for DTO style Input types, allowing for modifications of input data before being "set" onto DTO properties.  Of course, these modifications could be done in operation controllers before being persisted or passed to domain models.  Regardless, there are situations where it's ideal to modify via a setter in the DTO.  We've been using constructors for this purpose since the Input type resolver instantiates the DTO, passing the requisite constructor arguments, prior to hydrating the DTO properties.  It'd be cleaner to handle some of this in setters and offers a location for some custom validation as well - that's the big win.

**Comment by Lappihuan** [comment]
> Also, any thoughts on the Field annotation being used for input and output types? I'm a little concerned about the overall understanding on how these annotations work together with each other and the lack of clarity with their naming. In your example, you're going to get an input and output type with fields being mapped accordingly. However, to a noob, it's not inherently clear how this will happen.

This is exactly on of the edge cases. We are making the assumption that methods that start with `set` can only be InputFields and are filtered out for QueryFields.

While this works fine and seems to have no side effects it feels a tad bit hacky.

On the other hand i think PropertyAccessors like the one from Symfony have to make similar assumptions.
The one used in this lib does so too for example for getters:
https://github.com/thecodingmachine/graphqlite/blob/b69ba7e9d48fc91ef29504d93986bdb94d49d50a/src/Utils/PropertyAccessor.php#L27
I will check if i can reuse this.

**Comment by oojacoboo** [comment]
@Lappihuan I'm wondering if we could enforce the use of the `for` option on the `Field` attribute if the definition is ambiguous.

**Comment by Lappihuan** [comment]
@oojacoboo forcing `for` is not backwards compatible. but keep in mind, the field is not named `setLang` its named `lang` for input and output, its how getters worked since the beginning in this lib.
My concern was more with the technical implementation of it reather than the implicit separation, which should be natural to anyone familiar with graphqlite.
For a beginner, its a mather of having good docs explaining this very thing in a straight forward way.

**Comment by Lappihuan** [comment]
> Could you please outline the changes and improvements here overall? It'd be nice to have a list of functionality changes.
Setters should now support everything getters do:
- parameter middlewares:
    - Autowire
    - InjectUser
    - Any user defined parameter middlewares
- field middlewares:
    - Security
    - Right
    - Logged
    - Support for Validation/Assertation is given but needs to be added to those packages
    - user defined field middlewares are not supported out of the box, but can easily be added by implementing the `InputFieldMiddlewareInterface`
- `inputType` property is now able to overwrite the `update` property as described in #371
- potential fix for crash with circular reference in input types #449, currently building a testcase to confirm.

**Comment by oojacoboo** [comment]
@Lappihuan I've added a failing test, `@group PR-466`.  I came across the error while testing this PR on a rather large app.

**Comment by Lappihuan** [comment]
@oojacoboo i fixed a minor issue where the inputFields property was was not initialized as empty array.
Was this the bug you are having?

**Comment by oojacoboo** [comment]
@Lappihuan Yea, I guess the field annotations should be moved to the setters.  The issue is what’s being passed to the constructor isn’t correct.

**Comment by Lappihuan** [comment]
@oojacoboo the field annotation is fine on properties, as long as there exists a public setter method.
I removed the exceptions from the setters and the test is not failing now.

not sure whats you see wrong with the constructor

**Comment by oojacoboo** [comment]
Copied that over from another test, I wasn't able to test that it was properly failing.  Looks like it'll have to be adjusted.  We're getting the following error on our unit tests

```
TypeError: Foo\GraphQL\Operation\Employment\InputType\Employment::__construct(): Argument #1 ($startDate) must be of type ?Foo\Date, string given
```

`Foo\Date` is resolved with a custom root type mapper.

**Comment by Lappihuan** [comment]
@oojacoboo constructor hydration proves to be challenging in the way it does not integrate with the whole middleware pattern.

The middlewares can go and wrap a field with other logic like for example Security, or any other user defined middleware. Now in those middlewares need to have access to the resolver and from those resolvers you are always supposed to get the DTO you are working with. In the case of the Security middleware you can use it to access properties on the DTO in the expression via `this` so if you have a `secret` Field on the DTO you can check it in the expression via `this.secret == "Foo"`. 

This works well in the case of setting properties directly or via setters since the DTO is already instantiated. But with Fields that are supposed be hydrated via constructor this whole system does not make sense. 🐔🥚

I feel like constructor hydration has to be a second class citizen and not support any middlewares.

I will try and come up with a implementation that determines if a field is hydrated via constructor and if so it will skip all the fancy FieldDescriptor and middleware stuff and just be used to instantiate the DTO before its passed to hydration via setters and properties.

Any toughs?

**Comment by oojacoboo** [comment]
@Lappihuan I agree with skipping all the fancy business with the constructor.  Let's just get it instantiating and then we can do all the other middleware on the other fields being hydrated.  I'd be nice to check for fields being hydrated via the constructor to see if they're implementing any middleware logic, and if so, throw an exception.

There is one other option - fully constructing and hydrating the object and then post-processing all the middleware.  I suspect that this isn't worth it and would require recursively resolving all the fields in 2 passes - performance hit.

**Comment by Lappihuan** [comment]
>There is one other option - fully constructing and hydrating the object and then post-processing all the middleware. I suspect that this isn't worth it and would require recursively resolving all the fields in 2 passes - performance hit.

Since i already do a second pass for setters and properties this wouldn't be a big deal.
But it assumes there is a way to do so. If the constructor is the only way to hydrate it would not work as expected.

I think is should be able to throw if middleware annotations are detected for constructor hydrated fields. :+1: 

Working on it right now.

**Comment by oojacoboo** [comment]
@Lappihuan if you're already having to do 2 passes, using reflection, you can check which fields are being hydrated and simply re-hydrate them...  I don't love it, but it would retain functionality and I don't see any real issues with it.

**Comment by Lappihuan** [comment]
@oojacoboo now i only do one pass for all constructor fields and one for setters/property fields. I detect early if the property is hydrated via constructor and if so check if middleware annotations are used if so i throw a exception explaining its not supported, else i just skip all the middleware stuff and resolve the properties for instantiation, after instantiation the new stuff kicks in. fields that are used in the constructor are pretty close to how they are handled in the lib right now.

**Comment by oojacoboo** [comment]
@Lappihuan I'll give it a test again ASAP.

File: src/FieldsBuilder.php [hunk 1]
```diff
@@ -824,4 +853,220 @@ private function getPrefetchMethodInfo(ReflectionClass $refClass, $reflector, ob
...
+                $inputFieldDescriptor = new InputFieldDescriptor();
+                $inputFieldDescriptor->setRefMethod($refMethod);
+                $inputFieldDescriptor->setIsUpdate($isUpdate);
+                $inputFieldDescriptor->setName($name);
+                $inputFieldDescriptor->setComment(trim($description));
+
+                $parameters = $refMethod->getParameters();
+                if ($injectSource === true) {
+                    $firstParameter = array_shift($parameters);
+                    // TODO: check that $first_parameter type is correct.
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:899]: Is this needed?  If not, let's update the code comment to something that's more helpful for other devs.

File: src/FieldsBuilder.php [hunk 2]
```diff
@@ -342,8 +370,10 @@ private function getFieldsByMethodAnnotations($controller, ReflectionClass $refC
         $annotations = $this->annotationReader->getMethodAnnotations($refMethod, $annotationName);
         foreach ($annotations as $queryAnnotation) {
             $description = null;
+            $methodName = $refMethod->getName();
 
             if ($queryAnnotation instanceof Field) {
+                if (strpos($methodName, 'set') === 0) continue;
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:376]: It'd be good to stick with block statements

File: src/FieldsBuilder.php [hunk 3]
```diff
@@ -824,4 +853,220 @@ private function getPrefetchMethodInfo(ReflectionClass $refClass, $reflector, ob
...
+            if ($fieldAnnotations instanceof Field) {
+                $for = $fieldAnnotations->getFor();
+                if ($typeName && $for && !in_array($typeName, $for)) {
+                    continue;
+                }
+                $description = $fieldAnnotations->getDescription();
+
+                $docBlockObj = $this->cachedDocBlockFactory->getDocBlock($refMethod);
+                $methodName = $refMethod->getName();
+                if (strpos($methodName, 'set') !== 0) continue;
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:884]: It'd be good to stick with block statements

File: src/FieldsBuilder.php [hunk 4]
```diff
@@ -824,4 +853,220 @@ private function getPrefetchMethodInfo(ReflectionClass $refClass, $reflector, ob
...
+                $inputFieldDescriptor->setMiddlewareAnnotations($this->annotationReader->getMiddlewareAnnotations($refMethod));
+
+                $field = $this->inputFieldMiddleware->process($inputFieldDescriptor, new class implements InputFieldHandlerInterface {
+                    public function handle(InputFieldDescriptor $inputFieldDescriptor): ?InputField
+                    {
+                        return InputField::fromFieldDescriptor($inputFieldDescriptor);
+                    }
+                });
+
+
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:941]: Unnecessary blank lines

File: src/FieldsBuilder.php [hunk 5]
```diff
@@ -824,4 +853,220 @@ private function getPrefetchMethodInfo(ReflectionClass $refClass, $reflector, ob
...
+            if ($annotation instanceof Field) {
+                $for = $annotation->getFor();
+                if ($typeName && $for && !in_array($typeName, $for)) {
+                    continue;
+                }
+
+                $description = $annotation->getDescription();
+
+
+                $name = $annotation->getName() ?: $refProperty->getName();
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:984]: Wouldn't the null coalescing operator be more appropriate here?  `??`

File: src/FieldsBuilder.php [hunk 6]
```diff
@@ -824,4 +853,220 @@ private function getPrefetchMethodInfo(ReflectionClass $refClass, $reflector, ob
...
+
+                $inputType = $annotation->getInputType();
+                $constructerParameters = $this->getClassConstructParameterNames($refClass);
+                $inputProperty = $this->typeMapper->mapInputProperty($refProperty, $docBlock, $name, $inputType, $defaultProperties[$refProperty->getName()] ?? null, $isUpdate ? true : null);
+
+                if (! $description) {
+                    $description = $inputProperty->getDescription();
+                }
+
+                if(in_array($name, $constructerParameters)) {
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:994]: Space after `if`... `if (`

File: src/FieldsBuilder.php [hunk 7]
```diff
@@ -824,4 +853,220 @@ private function getPrefetchMethodInfo(ReflectionClass $refClass, $reflector, ob
...
+                    $middlewareAnnotations = $this->annotationReader->getPropertyAnnotations($refProperty, MiddlewareAnnotationInterface::class);
+                    if($middlewareAnnotations !== []){
```
Comment (oojacoboo) [inline src/FieldsBuilder.php:996]: Space after `if`.  Not sure why phpcs isn't picking up on this.

File: src/InputField.php [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            'name' => $name,
+            'type' => $type,
+            'description' => $comment
+        ];
+
+        if (!(!$hasDefaultValue || $isUpdate)) {
+            $config['defaultValue'] = $defaultValue;
+        }
+
+        if($originalResolver !== null && $resolver !== null){
```
Comment (oojacoboo) [inline src/InputField.php:58]: Space after `if`

File: src/InputField.php [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            $fieldDescriptor->isUpdate(),
+            $fieldDescriptor->hasDefaultValue(),
+            $fieldDescriptor->getDefaultValue(),
+        );
+    }
+
+    public static function fromFieldDescriptor(InputFieldDescriptor $fieldDescriptor): self
+    {
+        $arguments = $fieldDescriptor->getParameters();
+//        if ($fieldDescriptor->getPrefetchMethodName() !== null) {
```
Comment (oojacoboo) [inline src/InputField.php:178]: Unused code should be removed

File: src/InputFieldDescriptor.php [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+
+    /**
+     * @param mixed|null $defaultValue
+     */
+    public function setDefaultValue($defaultValue): void
+    {
+        $this->defaultValue = $defaultValue;
+    }
+
+
```
Comment (oojacoboo) [inline src/InputFieldDescriptor.php:123]: Extra lines

File: src/InputFieldDescriptor.php [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+     */
+    public function setCallable(callable $callable): void
+    {
+        if ($this->originalResolver !== null) {
+            throw new GraphQLRuntimeException('You cannot modify the callable via setCallable because it was already used. You can still wrap the callable using getResolver/setResolver');
+        }
+        $this->callable = $callable;
+        $this->targetMethodOnSource = null;
+        $this->targetPropertyOnSource = null;
+//        $this->magicProperty = null;
```
Comment (oojacoboo) [inline src/InputFieldDescriptor.php:206]: Remove unused code

File: src/InputFieldDescriptor.php [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+
+    public function setTargetMethodOnSource(string $targetMethodOnSource): void
+    {
+        if ($this->originalResolver !== null) {
+            throw new GraphQLRuntimeException('You cannot modify the target method via setTargetMethodOnSource because it was already used. You can still wrap the callable using getResolver/setResolver');
+        }
+        $this->callable = null;
+        $this->targetMethodOnSource = $targetMethodOnSource;
+        $this->targetPropertyOnSource = null;
+//        $this->magicProperty = null;
```
Comment (oojacoboo) [inline src/InputFieldDescriptor.php:217]: Remove unused code

File: src/InputFieldDescriptor.php [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+        if (is_callable($this->callable)) {
+            /** @var callable&array{0:object, 1:string} $callable */
+            $callable = $this->callable;
+            $this->originalResolver = new ServiceResolver($callable);
+        } elseif ($this->targetMethodOnSource !== null) {
+            $this->originalResolver = new SourceResolver($this->targetMethodOnSource);
+        } elseif ($this->targetPropertyOnSource !== null) {
+            $this->originalResolver = new SourceInputPropertyResolver($this->targetPropertyOnSource);
+//        } elseif ($this->magicProperty !== null) {
```
Comment (oojacoboo) [inline src/InputFieldDescriptor.php:319]: Unused code

File: src/Middlewares/AuthorizationInputFieldMiddleware.php [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    public function process(InputFieldDescriptor $inputFieldDescriptor, InputFieldHandlerInterface $inputFieldHandler): ?InputField
+    {
+        $annotations = $inputFieldDescriptor->getMiddlewareAnnotations();
+
+        $loggedAnnotation = $annotations->getAnnotationByType(Logged::class);
+        assert($loggedAnnotation === null || $loggedAnnotation instanceof Logged);
+        $rightAnnotation = $annotations->getAnnotationByType(Right::class);
+        assert($rightAnnotation === null || $rightAnnotation instanceof Right);
+
+//        $failWith = $annotations->getAnnotationByType(FailWith::class);
```
Comment (oojacoboo) [inline src/Middlewares/AuthorizationInputFieldMiddleware.php:47]: Unused code

File: src/Middlewares/AuthorizationInputFieldMiddleware.php [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+//            Assert::isInstanceOf($type, OutputType::class);
+//            $inputFieldDescriptor->setType($type);
+//        }
+        if ($this->isAuthorized($loggedAnnotation, $rightAnnotation)) {
+            return $inputFieldHandler->handle($inputFieldDescriptor);
+        }
+
+        $hideIfUnauthorized = $annotations->getAnnotationByType(HideIfUnauthorized::class);
+        assert($hideIfUnauthorized instanceof HideIfUnauthorized || $hideIfUnauthorized === null);
+
```
Comment (oojacoboo) [inline src/Middlewares/AuthorizationInputFieldMiddleware.php:63]: More unused code

File: src/Middlewares/InputFieldHandlerInterface.php [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+declare(strict_types=1);
+
+namespace TheCodingMachine\GraphQLite\Middlewares;
+
+use TheCodingMachine\GraphQLite\InputField;
+use TheCodingMachine\GraphQLite\InputFieldDescriptor;
+
+/**
+ * @unstable See https://graphqlite.thecodingmachine.io/docs/semver.html
```
Comment (oojacoboo) [inline src/Middlewares/InputFieldHandlerInterface.php:11]: What is meant by this annotation?  Link is broken.

File: src/Middlewares/InputFieldMiddlewareInterface.php [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace TheCodingMachine\GraphQLite\Middlewares;
+
+use TheCodingMachine\GraphQLite\InputField;
+use TheCodingMachine\GraphQLite\InputFieldDescriptor;
+
+/**
+ * A middleware use to process annotations when evaluating a input field
+ *
+ * @unstable See https://graphqlite.thecodingmachine.io/docs/semver.html
```
Comment (oojacoboo) [inline src/Middlewares/InputFieldMiddlewareInterface.php:13]: What is meant by this annotation? Link is broken.

File: src/Middlewares/SecurityInputFieldMiddleware.php [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    /*private $logger;*/
+    /** @var AuthorizationServiceInterface */
+    private $authorizationService;
+
+    public function __construct(ExpressionLanguage $language, AuthenticationServiceInterface $authenticationService, AuthorizationServiceInterface $authorizationService/*, ?LoggerInterface $logger = null*/)
+    {
+        $this->language = $language;
+        $this->authenticationService = $authenticationService;
+        $this->authorizationService = $authorizationService;
+        /*$this->logger = $logger;*/
```
Comment (oojacoboo) [inline src/Middlewares/SecurityInputFieldMiddleware.php:41]: Unused code

File: src/Middlewares/SecurityInputFieldMiddleware.php [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        $annotations = $inputFieldDescriptor->getMiddlewareAnnotations();
+        /** @var Security[] $securityAnnotations */
+        $securityAnnotations = $annotations->getAnnotationsByType(Security::class);
+
+        if (empty($securityAnnotations)) {
+            return $inputFieldHandler->handle($inputFieldDescriptor);
+        }
+
+//        $failWith = $annotations->getAnnotationByType(FailWith::class);
```
Comment (oojacoboo) [inline src/Middlewares/SecurityInputFieldMiddleware.php:54]: Unused code

File: src/Middlewares/SecurityInputFieldMiddleware.php [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+            'authorizationService' => $this->authorizationService, // Used by the is_granted expression language function.
+            'authenticationService' => $this->authenticationService, // Used by the is_logged expression language function.
+            'this' => $callable->getObject(),
+        ];
+
+        $argsName = array_keys($parameters);
+        $argsByName = array_combine($argsName, $args);
+        Assert::isArray($argsByName);
+
+        /*if ($diff = array_intersect(array_keys($variables), array_keys($argsName))) {
```
Comment (oojacoboo) [inline src/Middlewares/SecurityInputFieldMiddleware.php:125]: Unused code

File: src/Types/InputType.php [hunk 21]
```diff
@@ -45,32 +45,25 @@ public function __construct(
...
-            foreach ($this->fields as $name => $field) {
-                $type = $field->getType();
+        $fields = function () use ($isUpdate, $inputName, $className, $fieldsBuilder) {
+            $inputFields = $fieldsBuilder->getInputFields($className, $inputName, $isUpdate);
 
-                if ($isUpdate && $type instanceof NonNull) {
-                    $type = $type->getWrappedType();
+            $fieldConfigs = [];
+            foreach($inputFields as $field){
+                if($field->forConstructorHydration()){
```
Comment (oojacoboo) [inline src/Types/InputType.php:54]: Space after `if`

File: src/Types/InputType.php [hunk 22]
```diff
@@ -45,32 +45,25 @@ public function __construct(
...
-            $fields = [];
-            foreach ($this->fields as $name => $field) {
-                $type = $field->getType();
+        $fields = function () use ($isUpdate, $inputName, $className, $fieldsBuilder) {
+            $inputFields = $fieldsBuilder->getInputFields($className, $inputName, $isUpdate);
 
-                if ($isUpdate && $type instanceof NonNull) {
-                    $type = $type->getWrappedType();
+            $fieldConfigs = [];
+            foreach($inputFields as $field){
```
Comment (oojacoboo) [inline src/Types/InputType.php:53]: Space after `foreach`

File: src/Types/InputType.php [hunk 23]
```diff
@@ -88,27 +81,30 @@ public function __construct(
...
     {
-        $mappedValues = [];
-        foreach ($this->fields as $field) {
-            $name = $field->getName();
-            if (! array_key_exists($name, $args)) {
+        $constructorArgs = [];
+        foreach ($this->constructorInputFields as $constructorInputField) {
+            $name = $constructorInputField->name;
+            $resolve = $constructorInputField->getResolve();
+            if(!array_key_exists($name, $args)) {
```
Comment (oojacoboo) [inline src/Types/InputType.php:88]: Space after `if`

File: tests/Fixtures/CircularInputReference/Controllers/CircularController.php [hunk 24]
```diff
@@ -0,0 +1,20 @@
+<?php
+
```
Comment (oojacoboo) [inline tests/Fixtures/CircularInputReference/Controllers/CircularController.php:2]: Superfluous lines

File: tests/Fixtures/CircularInputReference/Controllers/CircularController.php [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace TheCodingMachine\GraphQLite\Fixtures\CircularInputReference\Controllers;
+
+
```
Comment (oojacoboo) [inline tests/Fixtures/CircularInputReference/Controllers/CircularController.php:6]: Superfluous lines

File: tests/Fixtures/CircularInputReference/Types/CircularInputA.php [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+ */
+class CircularInputA
+{
+
+    /**
+     * @Field(inputType="CircularInputBInput")
+     * @var CircularInputB
+     */
+    private $circularInputB;
+
```
Comment (oojacoboo) [inline tests/Fixtures/CircularInputReference/Types/CircularInputA.php:19]: Superfluous lines

File: tests/Fixtures/CircularInputReference/Types/CircularInputB.php [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+ */
+class CircularInputB
+{
+
+    /**
+     * @Field(inputType="CircularInputAInput")
+     * @var CircularInputA
+     */
+    private $circularInputA;
+
```
Comment (oojacoboo) [inline tests/Fixtures/CircularInputReference/Types/CircularInputB.php:19]: Superfluous lines

**Review by oojacoboo — CHANGES_REQUESTED** [review]
@Lappihuan good news, I've had a chance to test the latest code with this PR on a large application - all tests are passing!

I've added a code review here.  We probably should focus on getting `phpcs` rules updated to catch a lot of these code style comments I've made, then get that passing.

Looking forward to getting this merged in.  Thanks again for all the work on this!

File: src/FieldsBuilder.php (hunk 1 above)
Comment (Lappihuan) [reply src/FieldsBuilder.php:899]: this is copied from the getter part of the fields builder which has this comment, from what i understand, there should be a sanity check if the injected source is actually the same Type as hinted.

I'd leave it for now.

File: src/FieldsBuilder.php (hunk 5 above)
Comment (Lappihuan) [reply src/FieldsBuilder.php:984]: its also just copied from the equivalent code for output types, might have been a backwards compatible reason for it.

i think it should probably be updated in the whole library at some point, maybe using rector to automate it.

File: src/InputFieldDescriptor.php (hunk 11 above)
Comment (Lappihuan) [reply src/InputFieldDescriptor.php:206]: the magic property code is to be enabled in a future PR

File: src/InputFieldDescriptor.php (hunk 12 above)
Comment (Lappihuan) [reply src/InputFieldDescriptor.php:217]: the magic property code is to be enabled in a future PR

File: src/InputFieldDescriptor.php [hunk 28]
```diff
@@ -0,0 +1,345 @@
...
+
+    public function setTargetPropertyOnSource(?string $targetPropertyOnSource): void
+    {
+        if ($this->originalResolver !== null) {
+            throw new GraphQLRuntimeException('You cannot modify the target method via setTargetMethodOnSource because it was already used. You can still wrap the callable using getResolver/setResolver');
+        }
+        $this->callable = null;
+        $this->targetMethodOnSource = null;
+        $this->targetPropertyOnSource = $targetPropertyOnSource;
+//        $this->magicProperty = null;
```
[the remark this answers is not in the document]
Comment (Lappihuan) [reply src/InputFieldDescriptor.php:228]: the magic property code is to be enabled in a future PR

File: src/InputFieldDescriptor.php (hunk 13 above)
Comment (Lappihuan) [reply src/InputFieldDescriptor.php:319]: the magic property code is to be enabled in a future PR

File: src/Middlewares/InputFieldHandlerInterface.php (hunk 16 above)
Comment (Lappihuan) [reply src/Middlewares/InputFieldHandlerInterface.php:11]: copied from FieldHandlerInterface

File: src/Middlewares/InputFieldMiddlewareInterface.php (hunk 17 above)
Comment (Lappihuan) [reply src/Middlewares/InputFieldMiddlewareInterface.php:13]: copied from FieldMiddlewareInterface

File: src/Middlewares/SecurityInputFieldMiddleware.php [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+    /** @var ExpressionLanguage */
+    private $language;
+    /** @var AuthenticationServiceInterface */
+    private $authenticationService;
+    /** @var LoggerInterface|null */
+    /*private $logger;*/
+    /** @var AuthorizationServiceInterface */
+    private $authorizationService;
+
+    public function __construct(ExpressionLanguage $language, AuthenticationServiceInterface $authenticationService, AuthorizationServiceInterface $authorizationService/*, ?LoggerInterface $logger = null*/)
```
[the remark this answers is not in the document]
Comment (Lappihuan) [reply src/Middlewares/SecurityInputFieldMiddleware.php:36]: this is copied from the original SecurityFieldMiddleware which has a todo to add Logger support.
i assume this is preparation for that

File: src/Middlewares/SecurityInputFieldMiddleware.php (hunk 18 above)
Comment (Lappihuan) [reply src/Middlewares/SecurityInputFieldMiddleware.php:41]: connected to the logger stuff above

**Comment by Lappihuan** [comment]
@oojacoboo let me know if there is something to change on the unresolved points

File: src/InputFieldDescriptor.php (hunk 11 above)
Comment (oojacoboo) [reply src/InputFieldDescriptor.php:172]: Let's add a comment with any commented out code.  In general, we should remove virtually all commented out code, but small lines like this are fine if we can share with others why it's there.

File: src/InputFieldDescriptor.php (hunk 12 above)
Comment (oojacoboo) [reply src/InputFieldDescriptor.php:183]: Let's add a comment with any commented out code. In general, we should remove virtually all commented out code, but small lines like this are fine if we can share with others why it's there.

File: src/InputFieldDescriptor.php (hunk 28 above)
[the remark this answers is not in the document]
Comment (oojacoboo) [reply src/InputFieldDescriptor.php:194]: Let's add a comment with any commented out code. In general, we should remove virtually all commented out code, but small lines like this are fine if we can share with others why it's there.

File: src/InputFieldDescriptor.php (hunk 13 above)
Comment (oojacoboo) [reply src/InputFieldDescriptor.php:275]: Let's add a comment with any commented out code. In general, we should remove virtually all commented out code, but small lines like this are fine if we can share with others why it's there.

File: src/Middlewares/InputFieldHandlerInterface.php (hunk 16 above)
Comment (oojacoboo) [reply src/Middlewares/InputFieldHandlerInterface.php:11]: Let's remove it then.  The link is dead and doesn't make sense.

File: src/Middlewares/InputFieldMiddlewareInterface.php (hunk 17 above)
Comment (oojacoboo) [reply src/Middlewares/InputFieldMiddlewareInterface.php:13]: Let's remove it then. The link is dead and doesn't make sense.

File: src/Middlewares/SecurityInputFieldMiddleware.php (hunk 29 above)
[the remark this answers is not in the document]
Comment (oojacoboo) [reply src/Middlewares/SecurityInputFieldMiddleware.php:36]: Let's add a comment for the reasoning for it being there.

File: src/Middlewares/SecurityInputFieldMiddleware.php (hunk 18 above)
Comment (oojacoboo) [reply src/Middlewares/SecurityInputFieldMiddleware.php:41]: Let's add a comment for the reasoning for it being there.

**Comment by oojacoboo** [comment]
@Lappihuan thanks, added some more comments.  In general, if we're leaving commented out code, we should at least have a comment saying why it's there so it doesn't cause any confusion.

**Review by oojacoboo — APPROVED** [review]

**Comment by oojacoboo** [comment]
@Lappihuan I've merged in this PR.  Documentation wasn't completed and I'm doing that now, in addition to versioning the docs for a 6.0 release.

**Pull request merged**